File: stock.js
const randomInt = (min, max) =>
  Math.floor(Math.random() * (max - min + 1)) + min;

const offerMargin = () => randomInt(0, 10);

const genArrayRandom = (length, min, max) =>
  Array.from({ length }, () => randomInt(min, max));

const genArray = (length, number) => Array.from({ length }, () => number);

const TIME = 150;
const NUMBER_OF_OFFERS = 10;

const market = {
  buying: genArrayRandom(NUMBER_OF_OFFERS, 4, 25),
  selling: genArrayRandom(NUMBER_OF_OFFERS, 25, 50),

  maxHistory: genArray(TIME, 0),
  minHistory: genArray(TIME, 1000),

  lastPrice: null,
  delta: null,

  buyPrice() {
    return Math.min(...this.selling);
  },

  buy(time) {
    const price = this.buyPrice();
    const position = this.selling.indexOf(price);
    this.selling = this.selling.filter((_, i) => i != position);
    this._exchange(price, time);
    this._new_sell_offer();
  },

  sell(time) {
    const price = this.sellPrice();
    const position = this.buying.indexOf(price);
    this.buying = this.buying.filter((_, i) => i != position);
    this._exchange(price, time);
    this._new_buy_offer();
  },

  sellPrice() {
    return Math.max(...this.buying);
  },

  _exchange(price, time) {
    if (this.lastPrice !== null) {
      this.delta = price - this.lastPrice;
    }
    this.lastPrice = price;
    this.minHistory[time] = Math.min(price, this.minHistory[time]);
    this.maxHistory[time] = Math.max(price, this.maxHistory[time]);
  },

  _new_buy_offer() {
    this.buying.push(Math.max(this.buyPrice() - offerMargin(), 1));
  },

  _new_sell_offer() {
    this.selling.push(offerMargin() + this.sellPrice());
  },
};

const intervalGraph = (market) => {
  const maxHistory = market.maxHistory;
  const minHistory = market.minHistory;
  const price = Math.max(...maxHistory, ...minHistory);
  const grid = [];
  for (let i = 0; i < price; i++) {
    grid.push([]);
    for (let j = 0; j < TIME; j++) {
      grid[i].push(" ");
    }
  }
  for (let t = 0; t < TIME; t++) {
    const minPrice = minHistory[t];
    const maxPrice = maxHistory[t];
    for (let p = minPrice; p < maxPrice; p++) {
      grid[price - p][t] = "|";
    }
    grid[price - minPrice][t] = "+";
    grid[price - maxPrice][t] = "+";
  }
  console.log(`\nprice(${price})`);
  for (let i = 0; i < grid.length; i++) {
    const row = grid[i];
    const stringRow = row.join("");
    console.log(" |", stringRow);
  }
  console.log(" |_" + "_".repeat(TIME) + `__ time(${TIME})`);
};

const m = market;
for (let t = 0; t < TIME; t++) {
  const delta = m.delta;
  if (Math.random() < 0.5) {
    m.buy(t);
  } else {
    m.sell(t);
  }
  if (delta === null) {
    continue;
  }
  if (delta < 0) {
    m.buy(t);
    m.buy(t);
  }
  if (delta > 0) {
    m.sell(t);
    m.sell(t);
  }
}

intervalGraph(m);

File: stock.rb
require 'securerandom'


TIME = 100
NUMBER_OF_ORDERS = 10


def random(from, to)
  SecureRandom.random_number(to - from) + from
end

def ask_bid_spread
  random(1, 12)
end


module Offering
  def initialize(orders)
    @orders = orders
  end

  def get_orders
    @orders
  end

  def add_order(order)
    @orders << order
  end

  def remove_order(order)
    index = @orders.index(order)
    if index.nil?
        raise "order doesn't exist somehow"
    else
        @orders.delete_at(index)
    end
  end

  def best_price
    raise "not implemented"
  end

  def new_limit_order(price)
    add_order(price)
  end

  def market_order(new_limit_order)
    price = best_price
    remove_order(price)
    add_order(new_limit_order)
    price
  end
end

class Asking
  include Offering

  def best_price
    @orders.min
  end
end

class Biding
  include Offering

  def best_price
    @orders.max
  end
end

class Market
  attr_reader :delta

  def initialize
    @biding = Biding.new((0...NUMBER_OF_ORDERS).map { random(4, 25) })
    @asking = Asking.new((0...NUMBER_OF_ORDERS).map { random(25, 50) })
    @max_history = Array.new(TIME, 0)
    @min_history = Array.new(TIME, 1000)
    @last_price = @biding.best_price
    @delta = 0
  end

  def buy_market_order(time)
    new_limit_order = [@biding.best_price - ask_bid_spread, 1].max
    price = @asking.market_order(new_limit_order)
    record_exchange(time, price)
  end

  def sell_market_order(time)
    new_limit_order = @asking.best_price + ask_bid_spread
    price = @biding.market_order(new_limit_order)
    record_exchange(time, price)
  end

  def record_exchange(time, price)
    @delta = price - @last_price
    @last_price = price
    @min_history[time] = [@min_history[time], price].min
    @max_history[time] = [@max_history[time], price].max
  end

  def interval_graph
    max_history = @max_history
    min_history = @min_history
    max_price = [max_history.max, min_history.max].max

    grid = Array.new(max_price + 1) { Array.new(TIME, '  ') }

    (0...TIME).each do |turn|
      min_price = min_history[turn]
      max_price = max_history[turn]

      (min_price...max_price).each do |price|
        grid[grid.length - price - 1][turn] = ' |'
      end
      grid[grid.length - min_price - 1][turn] = ' +'
      grid[grid.length - max_price - 1][turn] = ' +'
    end

    puts "\nprice(#{max_price})"

    grid.each do |row|
      puts " | #{row.join}"
    end

    puts " |_#{'_' * (2 * TIME)}__ time(#{TIME})"
  end
end


market = Market.new

TIME.times do |turn|
  delta = market.delta
  if SecureRandom.random_number < 0.5
    market.buy_market_order(turn)
  else
    market.sell_market_order(turn)
  end

  if delta < 0
    market.buy_market_order(turn)
    market.buy_market_order(turn)
  end

  if delta > 0
    market.sell_market_order(turn)
    market.sell_market_order(turn)
  end
end

market.interval_graph
